File: packages/python/chp_core/sealing.py
"""Sealed payloads — payload confidentiality over the evidence chain (chp-v0.2.md
§16, proposal 0025).

The sibling of selective disclosure (§14): because ``chp-event-hash-v2`` binds an
event's ``content_hash`` to ``payload_commitment = sha256(canon(plaintext))``, not
the inline payload, a payload can be *sealed* (replaced with an encrypted
``{chp_sealed}`` marker) exactly as it can be *withheld* — the chain, root, and
signature verify offline over the ciphertext with no key. Only the holder of the
recipient X25519 key unseals, then re-runs the §14 commitment check.

``chp-sealed-v1`` = ephemeral X25519 ECDH → HKDF-SHA256 (``info="chp-sealed-v1"``)
→ ChaCha20-Poly1305 over ``canon(plaintext)``. All primitives are in the installed
``cryptography`` library — no new dependency, no PyNaCl, no ed25519→x25519 map (the
recipient sealing key is a *separate* X25519 key, published as ``enc_public_key``
in the signed host attestation).
"""

from __future__ import annotations

import base64
import copy
import json
from pathlib import Path
from typing import Any, Callable

SEALED_SCHEME = "chp-sealed-v1"
SEALED_SCHEME_V2 = "chp-sealed-v2"  # multi-recipient envelope encryption (proposal 0030)
_HKDF_INFO = b"chp-sealed-v1"
_ENC_PRIVATE_NAME = "host_x25519"
_ENC_PUBLIC_NAME = "host_x25519.pub"

# The commitment binds the inline payload only under chp-event-hash-v2 (§14).
EVENT_HASH_V2 = "chp-event-hash-v2"


def _canon_bytes(payload: Any) -> bytes:
    """chp-stable-v1 canonical bytes of a payload — identical to the form
    ``_payload_commitment`` hashes, so an unsealed payload round-trips to the same
    commitment."""
    return json.dumps(payload if payload is not None else {}, sort_keys=True).encode()


# ── Recipient X25519 key management (a sibling of the ed25519 key dir) ───────


def generate_enc_keypair(key_dir: str | Path, *, overwrite: bool = False):
    """Generate the host's X25519 sealing keypair alongside its ed25519 identity.
    Writes ``host_x25519`` (raw base64 private) + ``host_x25519.pub`` (raw base64
    public). Returns the ``X25519PrivateKey``. The public key is published as
    ``enc_public_key`` in the host attestation so senders can seal to it."""
    from cryptography.hazmat.primitives import serialization
    from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey

    key_dir = Path(key_dir)
    key_dir.mkdir(parents=True, exist_ok=True)
    priv_path = key_dir / _ENC_PRIVATE_NAME
    pub_path = key_dir / _ENC_PUBLIC_NAME
    if priv_path.exists() and not overwrite:
        return load_enc_private_key(key_dir)
    priv = X25519PrivateKey.generate()
    raw_priv = priv.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption())
    raw_pub = priv.public_key().public_bytes(
        encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw)
    priv_path.write_bytes(base64.b64encode(raw_priv))
    pub_path.write_bytes(base64.b64encode(raw_pub))
    try:
        priv_path.chmod(0o600)
    except OSError:  # pragma: no cover - non-POSIX
        pass
    return priv


def load_enc_private_key(key_dir: str | Path):
    """Load the host's X25519 sealing private key, or None if unprovisioned."""
    from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey

    path = Path(key_dir) / _ENC_PRIVATE_NAME
    if not path.exists():
        return None
    return X25519PrivateKey.from_private_bytes(base64.b64decode(path.read_bytes()))


def load_enc_public_key_b64(key_dir: str | Path) -> str | None:
    """The host's X25519 sealing PUBLIC key (base64), or None if unprovisioned."""
    path = Path(key_dir) / _ENC_PUBLIC_NAME
    if not path.exists():
        return None
    return path.read_bytes().decode().strip()


# ── The chp-sealed-v1 envelope ──────────────────────────────────────────────


def _derive_key(shared: bytes) -> bytes:
    from cryptography.hazmat.primitives import hashes
    from cryptography.hazmat.primitives.kdf.hkdf import HKDF

    return HKDF(algorithm=hashes.SHA256(), length=32, salt=None,
                info=_HKDF_INFO).derive(shared)


def _seal_bytes(recipient_pub_b64: str, plaintext: bytes,
                *, _esk_seed: bytes | None = None, _nonce: bytes | None = None) -> dict:
    """Seal ``plaintext`` to a recipient X25519 public key → a ``chp-sealed-v1``
    envelope. Fresh ephemeral key + nonce per call (non-deterministic ciphertext).
    ``_esk_seed``/``_nonce`` are test/vector hooks for a reproducible envelope — a
    real seal MUST leave them None (a reused ephemeral key/nonce is unsafe)."""
    import os

    from cryptography.hazmat.primitives import serialization
    from cryptography.hazmat.primitives.asymmetric.x25519 import (
        X25519PrivateKey, X25519PublicKey)
    from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305

    recipient = X25519PublicKey.from_public_bytes(base64.b64decode(recipient_pub_b64))
    esk = (X25519PrivateKey.from_private_bytes(_esk_seed) if _esk_seed
           else X25519PrivateKey.generate())
    epk = esk.public_key().public_bytes(
        encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw)
    key = _derive_key(esk.exchange(recipient))
    nonce = _nonce if _nonce is not None else os.urandom(12)
    ct = ChaCha20Poly1305(key).encrypt(nonce, plaintext, None)
    return {
        "scheme": SEALED_SCHEME,
        "epk": base64.b64encode(epk).decode(),
        "nonce": base64.b64encode(nonce).decode(),
        "ct": base64.b64encode(ct).decode(),
    }


def _seal_bytes_multi(recipient_pub_b64s: list[str], plaintext: bytes, *,
                      _cek: bytes | None = None, _content_nonce: bytes | None = None,
                      _esk_seeds: list[bytes] | None = None,
                      _wrap_nonces: list[bytes] | None = None) -> dict:
    """Seal ``plaintext`` to N recipient X25519 public keys → a ``chp-sealed-v2``
    envelope (envelope encryption). A single random 32-byte content key encrypts
    the payload ONCE (one ``ct``); that key is then wrapped per recipient by
    reusing ``_seal_bytes`` (a v1 seal of the 32-byte key). Any one recipient key
    recovers the content key and decrypts. Test hooks (``_cek``/``_content_nonce``/
    ``_esk_seeds``/``_wrap_nonces``) reproduce a fixed envelope for vectors — a real
    seal MUST leave them None."""
    import os

    from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305

    cek = _cek if _cek is not None else os.urandom(32)
    content_nonce = _content_nonce if _content_nonce is not None else os.urandom(12)
    ct = ChaCha20Poly1305(cek).encrypt(content_nonce, plaintext, None)
    recipients = []
    for i, pub in enumerate(recipient_pub_b64s):
        wrap = _seal_bytes(pub, cek,
                           _esk_seed=(_esk_seeds[i] if _esk_seeds else None),
                           _nonce=(_wrap_nonces[i] if _wrap_nonces else None))
        # Drop the redundant per-recipient scheme (v2 parent implies the v1 wrap).
        recipients.append({"epk": wrap["epk"], "nonce": wrap["nonce"],
                           "wrapped_key": wrap["ct"]})
    return {
        "scheme": SEALED_SCHEME_V2,
        "nonce": base64.b64encode(content_nonce).decode(),
        "ct": base64.b64encode(ct).decode(),
        "recipients": recipients,
    }


def _unseal_bytes(envelope: dict, enc_private_key) -> bytes:
    """Recover the plaintext bytes from a ``chp-sealed-v1`` (single-recipient) or
    ``chp-sealed-v2`` (multi-recipient envelope encryption) envelope with the
    recipient X25519 private key. A wrong key or tampered ciphertext raises."""
    from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PublicKey
    from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305

    scheme = envelope.get("scheme")
    if scheme == SEALED_SCHEME:
        epk = X25519PublicKey.from_public_bytes(base64.b64decode(envelope["epk"]))
        key = _derive_key(enc_private_key.exchange(epk))
        return ChaCha20Poly1305(key).decrypt(
            base64.b64decode(envelope["nonce"]), base64.b64decode(envelope["ct"]), None)
    if scheme == SEALED_SCHEME_V2:
        cek = None
        for r in envelope.get("recipients") or []:
            try:  # trial-unwrap: reconstruct the v1 wrap of the content key
                cek = _unseal_bytes({"scheme": SEALED_SCHEME, "epk": r["epk"],
                                     "nonce": r["nonce"], "ct": r["wrapped_key"]},
                                    enc_private_key)
                break
            except Exception:
                continue  # not our wrap — try the next recipient
        if cek is None:
            raise ValueError("no recipient key unwraps this chp-sealed-v2 envelope")
        return ChaCha20Poly1305(cek).decrypt(
            base64.b64decode(envelope["nonce"]), base64.b64decode(envelope["ct"]), None)
    raise ValueError(f"unknown sealing scheme: {scheme!r}")


# ── Bundle-level seal / unseal (mirrors signing.withhold_payloads) ───────────


def seal_payloads(bundle: dict, recipient_enc_pubkey: str | list[str],
                  predicate: Callable[[dict], bool] | None = None) -> dict:
    """Return a copy of ``bundle`` with every selected chp-event-hash-v2 event's
    ``payload`` replaced by a ``{"chp_sealed": <envelope>}`` marker (§16). The
    ``payload_commitment`` and ``content_hash`` are untouched, so the root and the
    ORIGINAL signature still verify — no re-signing, no store mutation, exactly like
    ``withhold_payloads``. An already-withheld or already-sealed payload is left
    as-is.

    ``recipient_enc_pubkey`` is a single X25519 public key (``chp-sealed-v1``,
    byte-identical to proposal 0025) OR a **list** of keys → ``chp-sealed-v2``
    envelope encryption (proposal 0030): the payload is encrypted once and readable
    by ANY of the N recipients."""
    if predicate is None:
        predicate = lambda ev: True  # noqa: E731
    out = copy.deepcopy(bundle)
    for ev in out.get("events") or []:
        if ev.get("hash_scheme") != EVENT_HASH_V2 or not ev.get("payload_commitment"):
            continue
        payload = ev.get("payload")
        if isinstance(payload, dict) and ("chp_withheld" in payload or "chp_sealed" in payload):
            continue  # already withheld/sealed
        if predicate(ev):
            pt = _canon_bytes(payload)
            env = (_seal_bytes_multi(recipient_enc_pubkey, pt)
                   if isinstance(recipient_enc_pubkey, list)
                   else _seal_bytes(recipient_enc_pubkey, pt))
            ev["payload"] = {"chp_sealed": env}
    return out


def unseal_payload(marker: dict, enc_private_key) -> Any:
    """Recover the plaintext payload from a ``{"chp_sealed": <envelope>}`` marker
    with the recipient X25519 private key. Returns the payload object; a caller
    then re-runs the §14 commitment check to confirm integrity."""
    envelope = marker.get("chp_sealed") if isinstance(marker, dict) else None
    if not isinstance(envelope, dict):
        raise ValueError("not a sealed payload marker")
    return json.loads(_unseal_bytes(envelope, enc_private_key))


def unseal_bundle(bundle: dict, enc_private_key) -> dict:
    """Return a copy of ``bundle`` with every ``{chp_sealed}`` payload decrypted
    back to plaintext (the inverse of ``seal_payloads``). Only the recipient can."""
    out = copy.deepcopy(bundle)
    for ev in out.get("events") or []:
        payload = ev.get("payload")
        if isinstance(payload, dict) and "chp_sealed" in payload:
            ev["payload"] = unseal_payload(payload, enc_private_key)
    return out

File: packages/chp-sdk/src/sealing.ts
/**
 * Sealed payloads — payload confidentiality over the evidence chain (chp-v0.2.md
 * §16, proposal 0025). The sibling of selective disclosure: a payload is encrypted
 * to a recipient's X25519 key and replaced with a `{chp_sealed}` marker, keeping
 * its `payload_commitment` so the chain/root/signature verify offline with no key.
 *
 * `chp-sealed-v1` = ephemeral X25519 ECDH → HKDF-SHA256 (info="chp-sealed-v1") →
 * ChaCha20-Poly1305 over canon(plaintext). Byte-compatible with Python
 * `chp_core.sealing`; `node:crypto` only (no new dependency). X25519 raw keys are
 * DER-wrapped exactly as ed25519 keys are in signing.ts (OID 1.3.101.110).
 */

// Runtime-neutral: X25519 (RFC 7748) + HKDF-SHA256 (RFC 5869) + ChaCha20-Poly1305 (RFC 8439), all via
// @noble. Raw keys throughout — no DER wrapping. Byte-compatible with the previous node:crypto path
// (all three are standardized); sealing.test.ts gates it.
import { x25519 } from '@noble/curves/ed25519.js';
import { hkdf } from '@noble/hashes/hkdf.js';
import { sha256 } from '@noble/hashes/sha2.js';
import { chacha20poly1305 } from '@noble/ciphers/chacha.js';
import { randomBytes } from './crypto.js';
import { canon, type JsonValue } from './canon.js';

export const SEALED_SCHEME = 'chp-sealed-v1';
export const SEALED_SCHEME_V2 = 'chp-sealed-v2';  // multi-recipient (proposal 0030)
const HKDF_INFO = Buffer.from('chp-sealed-v1', 'utf8');

function canonBytes(payload: JsonValue): Buffer {
  // chp-stable-v1 for hashing = JSON.stringify with sorted keys — identical to the
  // form Python's payload commitment hashes (canon() here matches that).
  return Buffer.from(canon(payload ?? {}), 'utf8');
}

function deriveKey(shared: Uint8Array): Buffer {
  return Buffer.from(hkdf(sha256, shared, new Uint8Array(0), HKDF_INFO, 32));
}

export interface SealedEnvelope { scheme: string; epk: string; nonce: string; ct: string; }
export interface SealedWrap { epk: string; nonce: string; wrapped_key: string; }
export interface SealedEnvelopeV2 { scheme: string; nonce: string; ct: string; recipients: SealedWrap[]; }

/** Generate an X25519 keypair; returns raw 32-byte private + base64 public. */
export function generateEncKeypair(): { privateRaw: Buffer; publicB64: string } {
  const priv = Buffer.from(randomBytes(32));
  const pub = Buffer.from(x25519.getPublicKey(priv));
  return { privateRaw: priv, publicB64: pub.toString('base64') };
}

function sealBytes(recipientPubB64: string, plaintext: Buffer): SealedEnvelope {
  const recipient = Buffer.from(recipientPubB64, 'base64');
  const esk = generateEncKeypair();
  const key = deriveKey(x25519.getSharedSecret(esk.privateRaw, recipient));
  const nonce = Buffer.from(randomBytes(12));
  // noble AEAD returns ciphertext‖tag (matches the Python "tag appended to ct" convention).
  const ct = Buffer.from(chacha20poly1305(key, nonce).encrypt(plaintext));
  return {
    scheme: SEALED_SCHEME, epk: esk.publicB64,
    nonce: nonce.toString('base64'), ct: ct.toString('base64'),
  };
}

/** ChaCha20-Poly1305 decrypt with a raw symmetric key (ct‖tag; throws on auth failure). */
function chachaDecrypt(key: Buffer, nonceB64: string, ctB64: string): Buffer {
  return Buffer.from(
    chacha20poly1305(key, Buffer.from(nonceB64, 'base64')).decrypt(Buffer.from(ctB64, 'base64')),
  );
}

/** Seal to N recipients → chp-sealed-v2 envelope encryption (proposal 0030): one
 * content key encrypts the payload once, wrapped per recipient via a v1 seal. */
function sealBytesMulti(recipientPubB64s: string[], plaintext: Buffer): SealedEnvelopeV2 {
  const cek = Buffer.from(randomBytes(32));
  const nonce = Buffer.from(randomBytes(12));
  const ct = Buffer.from(chacha20poly1305(cek, nonce).encrypt(plaintext));
  const recipients = recipientPubB64s.map((pub) => {
    const w = sealBytes(pub, cek);
    return { epk: w.epk, nonce: w.nonce, wrapped_key: w.ct };
  });
  return { scheme: SEALED_SCHEME_V2, nonce: nonce.toString('base64'), ct: ct.toString('base64'), recipients };
}

function unsealBytes(env: SealedEnvelope | SealedEnvelopeV2, encPrivateRaw: Buffer): Buffer {
  if (env.scheme === SEALED_SCHEME) {
    const e = env as SealedEnvelope;
    const key = deriveKey(x25519.getSharedSecret(encPrivateRaw, Buffer.from(e.epk, 'base64')));
    return chachaDecrypt(key, e.nonce, e.ct);
  }
  if (env.scheme === SEALED_SCHEME_V2) {
    const e = env as SealedEnvelopeV2;
    let cek: Buffer | null = null;
    for (const r of e.recipients ?? []) {
      try {  // trial-unwrap the content key from this recipient's v1 wrap
        cek = unsealBytes({ scheme: SEALED_SCHEME, epk: r.epk, nonce: r.nonce, ct: r.wrapped_key }, encPrivateRaw);
        break;
      } catch { /* not our wrap — try the next */ }
    }
    if (!cek) throw new Error('no recipient key unwraps this chp-sealed-v2 envelope');
    return chachaDecrypt(cek, e.nonce, e.ct);
  }
  throw new Error(`unknown sealing scheme: ${(env as { scheme?: string }).scheme}`);
}

/** Seal selected chp-event-hash-v2 payloads → `{chp_sealed}` markers (mirrors withholdPayloads). */
export function sealPayloads(
  bundle: Record<string, JsonValue>, recipientEncPubKey: string | string[],
  predicate?: (ev: Record<string, JsonValue>) => boolean,
): Record<string, JsonValue> {
  const multi = Array.isArray(recipientEncPubKey);
  const out = JSON.parse(JSON.stringify(bundle)) as Record<string, JsonValue>;
  for (const ev of (out.events as Array<Record<string, JsonValue>>) ?? []) {
    if (ev.hash_scheme !== 'chp-event-hash-v2' || !ev.payload_commitment) continue;
    const p = ev.payload as Record<string, unknown> | undefined;
    if (p && ('chp_withheld' in p || 'chp_sealed' in p)) continue;
    if (!predicate || predicate(ev)) {
      const pt = canonBytes(ev.payload);
      const env = multi ? sealBytesMulti(recipientEncPubKey, pt) : sealBytes(recipientEncPubKey, pt);
      ev.payload = { chp_sealed: env as unknown as JsonValue };
    }
  }
  return out;
}

/** Decrypt one `{chp_sealed}` marker (v1 or v2) to its plaintext payload. */
export function unsealPayload(marker: Record<string, JsonValue>, encPrivateRaw: Buffer): JsonValue {
  const env = marker?.chp_sealed as unknown as (SealedEnvelope | SealedEnvelopeV2) | undefined;
  if (!env || typeof env !== 'object') throw new Error('not a sealed payload marker');
  return JSON.parse(unsealBytes(env, encPrivateRaw).toString('utf8'));
}

/** Decrypt every `{chp_sealed}` payload back to plaintext (inverse of sealPayloads). */
export function unsealBundle(bundle: Record<string, JsonValue>, encPrivateRaw: Buffer): Record<string, JsonValue> {
  const out = JSON.parse(JSON.stringify(bundle)) as Record<string, JsonValue>;
  for (const ev of (out.events as Array<Record<string, JsonValue>>) ?? []) {
    const p = ev.payload as Record<string, JsonValue> | undefined;
    if (p && 'chp_sealed' in p) ev.payload = unsealPayload(p, encPrivateRaw);
  }
  return out;
}
